get-poetry.py

```diff
@@ -589,23 +588,63 @@ def _make_lib(self, version):
             finally:
                 gz.close()
 
+    def _which_python(self):
+        """Decides which python executable we'll embed in the launcher script."""
+        allowed_executables = ["python", "python3"]
+        if WINDOWS:
+            allowed_executables += ["py.exe -3", "py.exe -2"]
+
+        # \d in regex ensures we can convert to int later
+        version_matcher = re.compile(r"^Python (?P<major>\d+)\.(?P<minor>\d+)\..+$")
+        fallback = None
+        for executable in allowed_executables:
```

Comment:
@sdispater - the fact that this *prefers* python 2 is really unfortunate :-(
Any chance it could look for the newest python first?

Reply:
`python` should link to your preferred python version. So it's absolutely understandable that poetry will pick up this one. If your `python` doesn't link to your preferred python version, please look up how you can change it in your distribution.

Reply:
I don't buy it. `/usr/bin/python` is python2 for compatibility. Changing this on released distro is wrong.

What's the point of having `python3 get-poetry.py` installing to `python2` ?

Reply:
I'm wondering how your comment comply with https://www.python.org/dev/peps/pep-0394 :)

Reply:
Where it violates PEP394?
> If the python command is installed, it is expected to invoke either the same version of Python as the python3 command or as the python2 command.

Reply:
@finswimmer - `python` is very often the system python, and no, you shouldn't be messing around with that given that doing so can cripple package managers like yum. A common pattern is to, instead, have `python3` be the more recent interpreter and so the choice of ordering here is really unfortunate. I don't see how the experience would be degraded for your described use case were the ordering to be reversed.

This, unfortunately, appears to be part of a continuing trend in poetry to enforce everything working with a thing called `python` that has to already match the exact version of python you need for you project and having to torture your whole operating system and environment to make that happen, rather than the more common, convenient and flexible approach of allowing the python binary to be used by a tool like poetry to be specified explicitly where it can't be sensibly inferred.

Reply:
@cjw296 We already tried defaulting to `python3` if it was available but it was breaking some systems so we reverted the changes.

There is nothing wrong with defaulting to `python` and if it's not the version that you want to use you have the option to use the `env use` command which purpose is exactly that or you create and activate a virtual environment and Poetry will automatically pick it up.

Reply:
@sdispater we're talking about the python running poetry itself, not the python running managed project. does `env use` select the interpreter running poetry itself ?

Reply:
@sdispater actually, the point is not fallbacking to `python` but simply install poetry for the interpreter running get-poetry. Just like `pip` installs a package for the interpreter running pip.

Reply:
So does `get-pip.py` by default.

Reply:
@bersace The point of the official installer is to pick up the currently activate Python executable so you can easily switch between Python versions without having to install Poetry for each version. If we were to to hardcode the executable used to install Poetry this would defeat this purpose.

Reply:
By currently activated Python, do you mean pyenv or poetry-managed project python interpreter in a virtualenv ?

Reply:
@sdispater - what kind of breaking were you seeing when looking for `python3` first? 
How would use use the `env use` command when running `get-poetry.py`?

@bersace - I'm afraid @sdispater may well means `pyenv`. It's such an unfortunate choice to recommend as it's a terrible tool in professionally managed environments where `pyenv` either fails completely because development headers aren't available, or it builds broken python binaries because development headers for "optional" modules like `zlib` and `ncurses` aren't present, or perhaps worse still builds its own python binaries rather than using system-provided builds that have been optimised for the current environment by a sysadmin or build team.